benchmark/smallbank-bench

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env node
+/* eslint-disable require-jsdoc */
+
+const fs = require('fs');
+const {fork} = require('child_process');
+
+const [
+  properties,
+  numberOfAccount,
+  numberOfProcess,
+  duration,
+  rampUpTime,
+  operations,
+] = parseArgv();
+
+let counter = 0;
+let errorCount = 0;
+let totalCounter = 0;
+let latencyTotal = 0;
+
+const start = Date.now();
+const end = start + duration * 1000 + rampUpTime * 1000;
+
+const workers = [];
+for (let i = 0; i < numberOfProcess; i++) {
+  const worker = fork('./smallbank-bench-process', [
+    start.toString(),
+    properties,
+    numberOfAccount.toString(),
+    rampUpTime.toString(),
+    operations.join(','),
+  ]);
+
+  worker.on('message', (m) => {
+    if (m.counter) {
+      counter++;
+    }
+
+    if (m.error) {
+      errorCount++;
+    }
+
+    if (m.totalCounter) {
+      totalCounter += m.totalCounter;
+    }
+
+    if (m.latencyTotal) {
+      latencyTotal += m.latencyTotal;
+    }
+  });
+
+  workers.push(worker);
+}
+
+let from = start;
+(async () => {
+  while (true) {
+    const to = Date.now();
+    if (to >= end) {
+      for (const w of workers) {
+        w.kill();
+      }
+      break;
+    }
+
+    console.log(`${(counter * 1000) / (to - from)} tps`);
+    counter = 0;
+    from = Date.now();
+    await sleep(1000);
+  }
+})().then(() => {
+  const tps = (totalCounter * 1000) / (duration * 1000);
+  const averageLatency = latencyTotal / totalCounter;
+
+  console.log(`TPS: ${tps}`);
+  console.log(`Average-Latency(ms): ${averageLatency}`);
+  console.log(`Error-Counts: ${errorCount}`);
+  process.exit(0);
+});
+
+function sleep(ms) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+function parseArgv() {
+  let properties = 'client.properties.json';
+  let numberOfAccount = 10000;
+  let numberOfProcess = 1;
+  let duration = 200;
+  let rampUpTime = 60;
+  let operations = [
+    'transact_savings',
+    'deposit_checking',
+    'send_payment',
+    'write_check',
+    'amalgamate',
+  ];
+
+  const argv = process.argv.slice(2);
+
+  for (let i = 0; i < argv.length; i++) {
+    switch (argv[i]) {
+      case '--properties':
+      case '--config':
+        if (argv[i + 1] && fs.existsSync(argv[i + 1])) {
+          properties = argv[i + 1];
+          i++;
+        } else {
+          usage();
+          process.exit(1);
+        }
+        break;
+
+      case '--num-accounts':
+        if (argv[i + 1] && !isNaN(argv[i + 1]) && parseInt(argv[i + 1]) > 0) {
+          numberOfAccount = parseInt(argv[i + 1]);
+          i++;
+        } else {
+          usage();
+          process.exit(1);
+        }
+        break;
+
+      case '--num-processes':
```

Comment:
Sorry, we should have discussed the way to do it.

I think just using multiple processes is too efficient since a process is heavy so I think we should mix processes and threads (asynchronous multi-tasking) to make it comparable to the Java version.

One way to do this is that we let users specify `concurrency` and make the tool automatically divide the concurrency into multiple processes and "threads".
For example, if 16 concurrency is given on 4-core machine, we can create 4 processes to match with the number cores and create 4 "threads" for each process.

What do you think?

Reply:
I see.
I didn't think about that.
So, a sub-process will be in charge of a few concurrencies.
Unfortunately, we couldn't use worker_threads to implement it otherwise it might be more similar to Java's implementation.
I will revise smallbank-bench and smallbank-loader (according to Jun's feedback) by this idea.

Reply:
Sorry, I know we don't have threads. I meant multitasking with async.

Reply:
> Sorry, we should have discussed the way to do it.
> 
> I think just using multiple processes is too efficient since a process is heavy so I think we should mix processes and threads (asynchronous multi-tasking) to make it comparable to the Java version.
> 
> One way to do this is that we let users specify `concurrency` and make the tool automatically divide the concurrency into multiple processes and "threads".
> For example, if 16 concurrency is given on 4-core machine, we can create 4 processes to match with the number cores and create 4 "threads" for each process.
> 
> What do you think?

@feeblefakie 
I revised the implementation of smallbank-bench.
First, I rename the option name back to `--num-concurrencies`.
smallbank-bench now divides the requested concurrencies by the number of CPU,
then it forks subprocesses as the same number as CPU.
Each forked subprocess uses async functions with the number of divided concurrencies to run the benchmark.

I also revised smallbank-loader.
I didn't use `fork + async` way for smallbank-loader because I think it doesn't matter that smallbank-loader does works too efficiently, so I used an easier way to revise it.

@jnmt 
Jun, thank you for your advice.
I used `child_process` to revise smallbank-loader.
Now it can do jobs more efficiently too.